enhancements/cluster-logging/performance-tuning.md

```diff
@@ -0,0 +1,241 @@
+
+This is more efficient than duplicating the data again in a vector disk buffer,
+but only works for sources that support acknowledgement and "at-least-once" reliable delivery.
+
+Examples of acknowledgement sources:
+- Kafka: Kafka protocol has acknowledgements and at-least-once delivery.
+- HTTP: HTTP response can be used  to implement at-least-once.
+  Not all HTTP clients do this, but REST clients for data streaming usually do.
+- **File**: Vector's persistent "position" file can be used like an "acknowledgement" for file sources.
+  Reading from the persisted position after restart is equivalent to at-least-once delivery.
+
+#### Delivery policy implementation
+
+`Fast` policy does not provide any buffering, acknowledgement or other special configuration.
+
+`Safe` policy: can be implemented in different ways for different source/sink combinations, use the most efficient available
+- persistent sink buffer: Always for all combos, the only choice if any source does not support end-to-end ack.
+- end-to-end ack: generally more efficent if supported by all sources connected to a given sink.
+  - may interact with other options (rotate_wait, drop_oldest)
+  - may be combined with persistent buffering to give extra reliability for sources that support it.
+
+``` pseudocode
+for each `Safe` output:
+  set sink.acknowledgement=true for attached sink(s).
+  for each source of each input of each pipeline to output:
+    if source can participate in acknowledgement:
+	  set source.acknowledgement=true
+	else
+	  set output.buffer=disk
+```
+
+Persistent buffering is done to compensate for a `source` but happens on the `output`:
+- acknowledgement will still be used for sources that support it.
+- adding a buffer for the weak sources doesn not interfere with acknowledgement.
+
+NOTE: The existing `outputs[].limit` changes blocking behaviour in vector.
+Rate limits still should be enforced for `Safe` delivery.
+Review the existing limit code when implementing delivery policy so that the two work together properly.
+
+### Risks and Mitigations
+
+- Added complexity to forwarder.
+- Support cost of customers abusing or misunderstanding the parameters. \
+  For `delivery` this is mitigated by having the forward compute a "correct" config.
+- Increased customer demand for help "sizing" the logging stack: setting resources and predicting performance.
+- No new security risks are expected.
+
+### Drawbacks
+None.
+## Design Details
+
+### Open Questions
+None.
+
+### Test Plan
+None yet.
+
+May need improved tests to measure throughput and loss under sustained load.
+
+### Graduation Criteria
+None.
+
+### Upgrade / Downgrade Strategy
+
+Default configuration is backwards compatible.
+
+### Version Skew Strategy
+None.
+
+### Operational Aspects of API Extensions
+
+Backwards compatible extension to `ClusterLogForwarder` CR.
+
+#### Failure Modes
+
+Should make existing failure modes more predictable and reliable.
+
+#### Support Procedures
+
+None.
```

Comment:
As mentioned on the arch call, I suggest to add instructions of even better custom/recorded metrics on how to troubleshoot which output/sink is slowing down any other output/sinks at the source when using acks.

Reply:
Noted in the enhancement, but no concrete metric design yet.